### ISSUE TITLE
Port twisted.internet.inotify (along with the private twisted.python._inotify module) to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8211]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8211|
|Type|enhancement|
|Created|2016-02-17 04:28:09Z|
|Branch|https://github.com/twisted/twisted/tree/8211-rodrigc-inotify-py3|

I don't agree with adiroiban's comments so am putting it back
for review.  encode() / decode() uses [sys.getfilesystemencoding()](https://docs.python.org/3/library/sys.html#sys.getfilesystemencoding) which
has meanings on different platforms.
Twisted should, by default, honor the defaults, and only override them if a user explicitly forces it.

Waiting for feedback from another developer.

<details><summary>Searchable metadata</summary>

```
trac-id__8211 8211
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8211_rodrigc_inotify_py3 8211-rodrigc-inotify-py3
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1455683289846422 1455683289846422
changetime__1468910883618262 1468910883618262
version__None None
owner__rodrigc rodrigc

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [46808]) Branching to inotify-py3-8211.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_2"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/300

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @rodrigc|
|-|-|

Hi,

Thanks for the changes. They look good.

Major comments:

1. The decoding of the name obtained from the inotify buffer without using an explicit encoding and without handling decoding errors could generate ugly bugs.

On Linux, file names are just bytes and they can use any encoding ... you can end up with a strange Japanese file on an `en_US.UTF-8` and then the Python implicit decoding might break.

I think that name should be kept in bytes format when creating the child.


Minor comments


1. In tests, instead of ```path.getContent().decode()``` please use an explicit encoding. ex ``` path.getContent().decode('utf-8') ```

Even if we are in the testing code, we should still use best practices and alwasy make sure that we are in the control of the encoding used by the decode/encode operations.
As long as the file is generated using the same LANG environment, the implicit encoding should do its job... but thing will break once encoding is done in an environment with one LANG and decoding in another env with another LANG value.

2. twisted.python._inotify should be an internal thing and not advertised in the release notes. Twisted uesrs should not care about this module and we should not invite them to look into and and maybe start using things from it.


----------

Please address my comments and if agree with them, update the code and merge.
Otherwise please put the ticket back for review.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)|@rodrigc removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_6"></a>|@Lukasa set owner to @rodrigc|
|-|-|

To address Adi's concern, I think it would be best to change `encode()`/`decode()` to explicitly use `sys.getfilesystemencoding()`. This indicates that it's not that we *forgot* to provide an encoding, but that we explicitly want the filesystem encoding. Other than that, I agree with rodrigc: while on Linux the FS is technically a bag of bytes, it doesn't seem unreasonable to assume that, when using inotify, you're getting the FS encoding.

With that said, we may want to consider providing an override. I'm also willing to suggest that we should defer to a third developer, maybe hawkowl or glyph, to resolve the disagreement here.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_7"></a>|@rodrigc set owner to @Lukasa|
|-|-|

I'm OK with using sys.getfilesystemencoding(), since it has specific well documented meanings on different platforms: https://docs.python.org/3/library/sys.html#sys.getfilesystemencoding

See updated patch.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_8"></a>|@Lukasa commented|
|-|-|

Sorry, I think I was not clear enough.

`getfilesystemencoding` works fine in, say, `filename.path.encode()`. It does *not* work well on things like `path.getContent().decode()`. It is appropriate only when handling filenames, not when handling file content. The patch does not fully address the problem, as a result. See, for example, line 94 in `test_inotify.py` for an example of the problem. In that instance, it's better just to pick an arbitrary encoding. It doesn't matter what it is, so long as we're consistent with ourselves.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_9"></a>|@Lukasa commented|
|-|-|

Conversely, you *haven't* changed the `decode` on line 253 of `inotify.py`, which is where I'd have expected the change.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)|@Lukasa set owner to @rodrigc|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

But why try to decode at this level at all?

If you have LANG=C and somehow another application has just created a file in unicode with UTF-8 encoding, the inotify high level code will never receive the notification as this code will break.

But in other circumstances, the code can break even with an UTF-8 lang.

Here is the implicit decode in action... I have used os.listdir('.') and not os.listdir(u'.') so that the results are bytes and not Unicode.

```
$ echo $LANG
en_US.UTF-8

$ python2 -c "import os; print os.listdir('.')"
['.bzrignore', 'check_pyflakes.py', 'som\xc8\x9b_file']


$ python2 -c "import os; [name.decode() for name in os.listdir('.')]"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc8 in position 3: ordinal not in range(128)
```
  


This is why I am arguing that this API should just generate bytes and end users can decode it as they want.

----------

This is Linux specific code, so OSX is not a problem here, but keep in mind that OSX is a bit "special" when it comes to UTF-8 for the file names :)

getfilesystemencoding() is giving you the correct response (the encoding), but it does not tell you the unicode normalisation form... and on OSX you need to apply an extra layer of NFC normalization.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_12"></a>|@Lukasa commented|
|-|-|

In this instance I'm not wedded to it: I'd be happy enough to keep the file in bytes mode. Adi does raise a good point.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_13"></a>|@rodrigc set owner to @Lukasa|
|-|-|

OK, I changed it in inotify.py as well as per your suggestion.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_14"></a>|@markrwilliams commented|
|-|-|

In Python 3 `bytes.decode` [defaults to UTF-8](https://docs.python.org/3/library/stdtypes.html#bytes.decode), while on Python 2 it uses whatever `sys.getdefaultencoding`.

That means `.decode()` is wrong if UTF-8 is the desired encoding.  But as Adi says and shows, it is incredibly problematic to decode paths on any non-Windows platform.

May I suggest that this patch use the `surrogateescape` [error handler](https://www.python.org/dev/peps/pep-0383/) when decoding the inotify buffer? [It will ensure the decoded path can be re-encoded even if it contains bytes that could not be decoded](https://docs.python.org/3/library/os.html#file-names-command-line-arguments-and-environment-variables).

I believe the only change necessary would be to change line 254 of the latest commit to read:

```python
name = nameBytes.decode(sys.getfilesystemencoding(), 'surrogateescape')
```

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_15"></a>|@markrwilliams commented|
|-|-|

> May I suggest that this patch use the surrogateescape ​error handler when decoding the inotify buffer?

...And of course `surrogateescape` hasn't been backported to Python 2.  I don't know how decoding can be done safely on Python 2.  Anyway, this seems like more of a `FilePath` issue than one related to this patch.  So don't let my comments prevent this from being merged.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_16"></a>|@Lukasa set owner to @rodrigc|
|-|-|

So I have one question inline in the pull request about the move to bytes mode internally. Other than that I'm happy with this patch, but I'd like that question resolved before we continue.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|GitHub set status to `closed`|
|-|-|

In changeset c81c502e401dd85725d45e79c0201b4e6e6e6a49
```
#!CommitTicketReference repository="" revision="c81c502e401dd85725d45e79c0201b4e6e6e6a49"
Merge pull request #300 from twisted/8211-rodrigc-inotify-py3

Author: rodrigc
Reviewer: Lukasa, adiroiban
Fixes: #8211

Port twisted.internet.inotify and twisted.python._inotify to Python 3
```
